reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+
+      return Promise.reject(this._storedError);
+    }
+
+    if (this._state === 'closed' && this._ownerReadableByteStream === undefined) {
+      return Promise.resolve(CreateIterResultObject(new ctor(view.buffer, view.byteOffset, 0), true));
+    }
+
+    const promise = new Promise((resolve, reject) => {
+      const req = {
+        resolve,
+        reject,
+        byteOffset: view.byteOffset,
+        byteLength: view.byteLength,
+        ctor,
+        elementSize
+      };
+      this._readIntoRequests.push(req);
+    });
+
+    PullFromReadableByteStreamInto(
+        this._ownerReadableByteStream, view.buffer, view.byteOffset, view.byteLength, elementSize);
+
+    return promise;
+  }
+
+  releaseLock() {
+    if (!IsReadableByteStreamByobReader(this)) {
+      throw new TypeError(
+          'ReadableByteStreamByobReader.prototype.releaseLock can only be used on a ReadableByteStreamByobReader');
+    }
+
+    if (this._ownerReadableByteStream === undefined) {
+      return;
+    }
+
+    if (this._readIntoRequests.length > 0) {
+      throw new TypeError('Tried to release a reader lock when that reader has pending read() calls un-settled');
+    }
+
+    if (this._ownerReadableByteStream._state === 'readable') {
+      CloseReadableByteStreamReader(this);
+    }
+  }
+}
+
+function CallPull(controller) {
+  const source = controller._underlyingByteSource;
+
+  const pullFunction = source['pull'];
+  if (pullFunction === undefined) {
+    return;
+  }
+
+  const stream = controller._controlledReadableByteStream;
+
+  if (typeof pullFunction !== 'function') {
+    DestroyReadableByteStreamController(controller);
+    ErrorReadableByteStream(stream, new TypeError('pull property of an underlying byte source must be a function'));
+    return;
+  }
+
+  controller._callPullOrPullIntoLaterIfNeeded = false;
+  controller._insideUnderlyingByteSource = true;
+
+  try {
+    pullFunction.apply(source);
+  } catch (e) {
+    DestoryReadableByteStreamController(controller);
+    if (stream._state === 'readable') {
+      ErrorReadableByteStream(stream, e);
+    }
+  }
+
+  controller._insideUnderlyingByteSource = false;
+}
+
+function CallPullInto(controller) {
+  const source = controller._underlyingByteSource;
+
+  const pullIntoFunction = source['pullInto'];
+  if (pullIntoFunction === undefined) {
+    return;
+  }
+
+  const stream = controller._controlledReadableByteStream;
+
+  if (typeof pullIntoFunction !== 'function') {
+    DestroyReadableByteStreamController(controller);
+    ErrorReadableByteStream(stream, new TypeError('pullInto property of an underlying byte source must be a function'));
+    return;
+  }
+
+  assert(controller._pendingPullIntos.length > 0);
+  const pullIntoDescriptor = controller._pendingPullIntos[0];
+
+  controller._callPullOrPullIntoLaterIfNeeded = false;
+  controller._insideUnderlyingByteSource = true;
+
+  try {
+    pullIntoFunction.apply(source, [pullIntoDescriptor.buffer,
```

Comment:
It's (buffer, offsetToWriteAt, sizeAllowedToWrite). The buffer we received from the consumer is `byteLength` long. It's already filled for `bytesFilled` bytes. So, byteLength - bytesFilled bytes are available for writing.

I'm fine with changing this to Uint8Array. I kept this ArrayBufferView free since it seems a little complex interface is acceptable for the controller. But yeah, I realized that this is just inconsistent with enqueue() which accepts ArrayBufferViews.